### ISSUE TITLE
title attribute used as type and does not resolve

### ISSUE THREAD

**Issue by simonschllng** [bcherny/json-schema-to-typescript#17]
I noticed that any property in a json schema which has a title does not get resolved. Instead, the title is used as a type.

Example:

        "users": {
          "type": "array",
          "title": "user id array",
          "description": "Array of authorized user ids.",
          "items": {
            "type": "string"
          }
        }

gets translated to:

    users?: UserIdArray;

which leads to a undefined type `UserIdArray`. Is that behavior expected? How do I get json-schema-to-typescript to define the referenced types as well?

For now I created a workaround by removing ` || rule.title` from [line 168](https://github.com/bcherny/json-schema-to-typescript/blob/master/src/index.ts#L168). This generates:

    users?: string[];

**Comment by bcherny** [comment]
This seems like a bug to me. @Speedy37 What was your intention here?

**Comment by Speedy37** [comment]
yes it is a bug

**Comment by Speedy37** [comment]
it should generate:

```ts
interface MyInterface {
  users?: UserIdArray;
}
type UserIdArray = string[];
```

**Comment by bcherny** [comment]
@Speedy37 Can you take a look when you have a chance? Otherwise I can look later this week. Or if @simonschllng wants to give it a shot, go for it!

**Comment by bcherny** [comment]
Thanks for the quick response @Speedy37!

**Comment by simonschllng** [comment]
@bcherny, it still isn't working as of version 1.2.2.

I created a minimal example based on my actual schema:

```
{
  "$schema": "http://example.com/schema#",
  "id": "my-test",

  "definitions": {
    "my_object": {
      "$schema": "http://example.com/schema#",
      "type": "object",
      "headerTemplate": "{{i}} - {{self.name}}",
      "properties": {
        "name": {
          "type": "string",
          "title": "Name",
          "description": "Name of the object"
        },
        "coordinates": {
          "$ref": "#/definitions/geo",
          "title": "The coordinates",
          "description": "Coordinates of the object",
          "format": "location"
        },
        "content": {
          "type": "object",
          "title": "Content object",
          "properties": {
            "content_property": {
              "type": "array",
              "title": "Property of content",
              "description": "This is an array of strings.",
              "uniqueItems": true,
              "items": {
                "type": "string",
                "description": "A string."
              }
            }
          }
        }
      }
    },
    "geo": {
      "$schema": "http://json-schema.org/draft-04/schema#",
      "description": "A geographical coordinate",
      "type": "object",
      "properties": {
          "latitude": { "type": "number" },
          "longitude": { "type": "number" }
      }
    }

  },


  "allOf": [ {"$ref": "#/definitions/my_object"} ]
}
```


This generates:

```
/** A geographical coordinate */
interface Geo {
  latitude?: number;
  longitude?: number;
  [k: string]: any;
}
interface MyObject {
  name?: Name; // Name of the object
  coordinates?: Geo; // A geographical coordinate
  content?: ContentObject;
  [k: string]: any;
}
type MyTest = MyObject;
```

**Comment by bcherny** [comment]
Please upgrade to 2.0.0.

**Comment by simonschllng** [comment]
2.0.0 did fail on my schema with

```
path.js:7
    throw new TypeError('Path must be a string. Received ' + inspect(path));
    ^

TypeError: Path must be a string. Received undefined
    at assertPath (path.js:7:11)
    at Object.resolve (path.js:1148:7)
    at new Compiler (/.../node_modules/json-schema-to-typescript/dist/index.js:279:27)
    at Object.compile (/.../node_modules/json-schema-to-typescript/dist/index.js:557:12)
```

**Comment by bcherny** [comment]
@simonschllng Looking at your usage in https://github.com/bcherny/json-schema-to-typescript/issues/19, you're missing the required filePath parameter when calling `compile()`. I've made this parameter optional in https://github.com/bcherny/json-schema-to-typescript/issues/20, and published it as 2.0.1.

If you install the latest and try again it will work fine for your use case.

Ironically, using either TS or a JSON-Schema for the json-schema-to-ts API would have prevented this improper use :)
